pkg/lexer/state.go

```diff
@@ -0,0 +1,380 @@
+// Copyright © 2021 Rak Laptudirm <raklaptudirm@gmail.com>
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package lexer
+
+import (
+	"errors"
+	"unicode"
+
+	"github.com/raklaptudirm/mash/pkg/token"
+)
+
+var ErrEOF = errors.New("unexpected EOF")
+
+// run starts lexing the source in l and closes the lexer's token channel
+// when it is done.
+func (l *lexer) run() {
+	lexBlock(l, eof, token.EOF)
+	close(l.Tokens)
+}
+
+func lexBlock(l *lexer, eob rune, tok token.TokenType) {
+	for {
+		r := l.peek()
+		switch {
+		case r == eob:
+			l.consume()
+			l.emit(tok)
+			return // block lexed
+
+		// ignore all space runes
+		case unicode.IsSpace(r):
+			consumeAllSpace(l)
+
+		case r == '#':
+			lexComment(l)
+
+		// command or statement
+		default:
+			if isAlphabet(r) {
+				consumeWord(l)
+
+				word := l.literal()
+				// statement starts with keyword
+				if token.IsKeyword(word) {
+					t := token.Lookup(word)
+					l.emit(t)
+					l.insertSemi = t.InsertSemi()
+
+					lexStmt(l, eob)
+					goto insertSemi
```

Comment:
Do we really need goto in this situation? It's the same as `l.emit(token.SEMICOLON); continue;` right? Why add an extra flow control sequence?

Reply:
```go
63  // commands don't start with a keyword
64  // TODO: cleanup  <---
62  l.rdOffset = l.offset4
```
See line `64` XD
I will fix it though.